src/toil/lib/ec2.py

```diff
@@ -258,56 +366,220 @@ def create_ondemand_instances(ec2, image_id, spec, num_instances=1):
+                           tags: Optional[Dict[str, str]] = None) -> str:
+    """
+    Creates a launch template with the given name for launching instances with the given parameters.
+    
+    We only ever use the default version of any launch template.
+
+    Internally calls https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html?highlight=create_launch_template#EC2.Client.create_launch_template
+    
+    :param tags: Tags, if given, are applied to the template itself, all instances, and all volumes.
+    :param user_data: non-base64-encoded user data to pass to the instances.
+    
+    
+    :return: the ID of the launch template.
+    
+    
+    """
+    log.info('Creating launch template for %s instances ... ', instance_type)
+    
+    if isinstance(user_data, str):
+        # Make sure we have bytes
+        user_data = user_data.encode('utf-8')
+        
+    # Then base64 and decode back to str.
+    user_data = b64encode(user_data).decode('utf-8')
+    
+    template = {'ImageId': image_id,
+                'KeyName': key_name,
+                'SecurityGroupIds': security_group_ids,
+                'InstanceType': instance_type,
+                'UserData': user_data,
+                'BlockDeviceMappings': block_device_map,
+                'SubnetId': subnet_id}
+               
+    if instance_profile_arn:
+        # We could just retry when we get an error because the ARN doesn't
+        # exist, but we might as well wait for it.
+        wait_until_instance_profile_arn_exists(instance_profile_arn)
+        
+        # Add it to the request
+        template['IamInstanceProfile'] = {'Arn': instance_profile_arn}
+        
+    if placement_az:
+        template['Placement'] = {'AvailabilityZone': placement_az}
+               
+    if tags:
+        # Tag everything when we make it.
+        # TODO: just generate tags data once?
+        template['TagSpecifications'] = [{'ResourceType': 'instance', 'Tags': [{'Key': k, 'Value': v} for k, v in tags.items()]},
+                                         {'ResourceType': 'volume', 'Tags': [{'Key': k, 'Value': v} for k, v in tags.items()]}]
+
+    request = {'LaunchTemplateData': prune(template),
+               'LaunchTemplateName': template_name}
+               
+    if tags:
+        request['TagSpecifications'] = [{'ResourceType': 'launch-template', 'Tags': [{'Key': k, 'Value': v} for k, v in tags.items()]}]
+               
+    return ec2_client.create_launch_template(**request)['LaunchTemplate']['LaunchTemplateId']
+
+@retry(intervals=[5, 5, 10, 20, 20, 20, 20], errors=INCONSISTENCY_ERRORS)
+def create_auto_scaling_group(autoscaling_client: BaseClient,
+                              asg_name: str,
+                              launch_template_id: str,
+                              vpc_subnets: List[str],
+                              min_size: int = 0,
+                              max_size: int = 0,
```

Comment:
No, 0 means empty as the default. I think we want to default to not letting you make any instances, rather than to letting you make infinite instances. It prevents a missing parameter from going very wrong and wasting all your money.

Reply:
I'll just make min and max size both required.